typescript/src/backoff.ts

```diff
@@ -0,0 +1,97 @@
+  return true
+}
+
+/**
+ * @callback ErrorMatcherFn A function that returns true if the passed error
+ *           matches and false otherwise. Used to determine if a given error is
+ *           eligible for retry in `withBackoffRetries`.
+ * @param {Error} error The error to check for eligibility.
+ * @return {boolean} True if the error matches, false otherwise.
+ */
+type ErrorMatcherFn = (err: unknown) => boolean
+
+/**
+ * @callback RetrierFn A function that can retry any function passed to it a
+ *           set number of times, returning its result when it succeeds. Created
+ *           using `backoffRetrier`.
+ * @param {function(): Promise<T>} fn The function to be retried.
+ * @return {Promise<T>}
+ */
+
+/**
+ * A function that is called with execution status messages.
+ */
+type ExecutionLoggerFn = (msg: string) => void
+
+/**
+ * Returns a retrier that can be passed a function to be retried `retries`
+ * number of times, with exponential backoff. The result will return the
+ * function's return value if no exceptions are thrown. It will only retry if
+ * the function throws an exception matched by `matcher`; {@see retryAll } can
+ * be used to retry no matter the exception, though this is not necessarily
+ * recommended in production.
+ *
+ * Example usage:
+ *
+ *      await url.get("https://example.com/") // may transiently fail
+ *      // Retries 3 times with exponential backoff, no matter what error is
+ *      // reported by `url.get`.
+ *      backoffRetrier(3)(async () => url.get("https://example.com"))
+ *      // Retries 3 times with exponential backoff, but only if the error
+ *      // message includes "server unavailable".
+ *      backoffRetrier(3, (_) => _.message.includes('server unavailable'))(
+ *        async () => url.get("https://example.com"))
+ *      )
+ *
+ * @param {number} retries The number of retries to perform before bubbling the
+ *        failure out.
+ * @param {ExecutionLoggerFn} logger A logger function to pass execution messages.
+ * @param {ErrorMatcherFn} [errorMatcher=retryAll] A matcher function that
+ *        receives the error when an exception is thrown, and returns true if
+ *        the error should lead to a retry. A false return will rethrow the
+ *        error and terminate the retry loop.
+ * @returns {RetrierFn<T>} A function that can retry any function.
+ */
+export function backoffRetrier<T>(
+  retries: number,
+  logger: ExecutionLoggerFn = console.debug,
+  errorMatcher: ErrorMatcherFn = retryAll
+) {
+  return async (fn: () => Promise<T>): Promise<T> => {
+    for (let attempt = 0; attempt < retries; attempt++) {
+      try {
+        logger(`making attempt number ${attempt}`)
+
+        return await fn()
+      } catch (error: unknown) {
+        if (!errorMatcher(error)) {
+          // If the matcher doesn't match this error, rethrow and stop.
+          throw error
+        }
+
+        const backoffMillis = Math.pow(2, attempt) * 1000
```

Comment:
Can we make the backoff multiplier configurable? While working on #514, I realized that `1000 ms` is often not enough for Electrum retries, especially when errors are caused by the node being out of sync and, for example, not yet seeing the requested transaction. We can use `1000 ms` by default but I'd love to pass anther value if needed.

Reply:
Can't we just add more retries if we need a longer backoff?

Reply:
We can, but in some cases, it may make more sense to wait a little longer than bombarding the target service with new requests. This change should not be a big deal but can add great flexibility.

Reply:
a20f04ada97c204b1b5a1e6e8b36a13c55819195